README.md
<img src="docs/images/open-lasertag-logo.svg" alt="Open Lasertag Logo" width="100"/>

# Open Laser Tag
> because Laz00rs!

An cheap and DIY open source lasertag system, build with an ESP-32. Lasertag works with infrared. The tagger is build with a infrared LED and a infrared receiver, through which the players can hit each other. This basic functionality is given by the tagger and can be tested with two or more taggers standalone. For more complex games the esp32 is paired via bluetooth with a android app (work in progress). With the android app and a internet connection it will be possible to track statistics during a game and create more complex games e.g. with GPS location data.

This project currently consists of three parts:
* ESP-32 Firmware (https://github.com/open-laser-tag/tagger-firmware)
* Android App (https://github.com/open-laser-tag/android-app)
* node.js Server (https://github.com/open-laser-tag/server)

Videos about Open Laser Tag:
* Lightning Talk from Chaos Commuinication Camp 2019 (German, English dub available): https://media.ccc.de/v/Camp2019-10380-lightning_talks#t=10043

## Developing
clone repository with **submodules**
`git clone git@github.com:open-laser-tag/tagger.git --recurse-submodules`
There are a lot of options for writing code for the esp32 and for flashing it e.g. Arduino IDE, VScode with Arduino extention, PlattformIO. All possibilities to develop with the esp32 are listed and described [here](https://github.com/espressif/arduino-esp32/).

When uploading to esp32, don't forget to press IO0 on it when it tries to connect to the esp32.
![Button IO0 on ESP32 board](https://raw.githubusercontent.com/wiki/open-laser-tag/tagger/nodemcu_esp32_buttonio0.jpg)

When you want to keep track of which firmware is flashed on your tagger, you can add [update_git_hash.sh](https://github.com/open-laser-tag/tagger/blob/master/update_git_hash.sh) as git hook to your local git repository. Just add a file named `post-commit` to `.git/hooks/` and add the line `./update_git_hash.sh` or run this script manually before flashing, when you are not using the makefile.

### Arduino IDE
Prepare your IDE like described and then open
`tagger.ino`

`chose ESP32 DevModule` as Board
set partition scheme to `Minimal SPIFFS (Large `

### VScode with Arduino extention
Your arduino.json could look like this
```
{
    "port": "/dev/ttyUSB0",
    "board": "esp32:esp32:esp32",
    "configuration": "PSRAM=disabled,PartitionScheme=min_spiffs,CPUFreq=240,FlashMode=qio,FlashFreq=80,FlashSize=4M,UploadSpeed=921600,DebugLevel=debug",
    "sketch": "tagger.ino",
    "output": "./build"
}
```

### PlatformIO - VScode
All necessery informations are already provided in platformio.ini. When your esp32 is connected via USB you can simply press upload.

### PlatformIO - CLI
When you have installed PlatformIO on your computer, you can just use the provided Makefile in this repository and run
`make upload` to download all dependencies, build the code, and upload it to your ESP-32 in one step.

### flashing with esptool.py and release binary
1. install [esptool.py](https://github.com/espressif/esptool) with their instructions
2. Download both .bin files from newest [release](https://github.com/open-laser-tag/tagger/releases/).
3. Flash ESP with `esptool.py --chip esp32 --port /dev/ttyUSB0 write_flash 0x10000 tagger_firmware.ino.bin 0x8000 tagger_firmware.ino.partitions.bin` (+ press Button IO0)

## Tagger Hardware
One tagger is made of: 1x ESP32, 1x IR LED [TSUS 5202](https://www.conrad.de/de/p/vishay-tsus-5202-cqw-13-ir-emitter-950-nm-15-5-mm-radial-bedrahtet-184551.html), 1x IR receiver [TSOP31238](https://www.segor.de/#Q=TSOP31238&M=1), 1x power bank, 1x micro usb cable, some RGB LEDS WS2812 or APA102, 1x [lense](https://www.ebay.de/itm/2x-Cardboard-Virtual-Reality-VR-BiConvex-Lenses-Only-25mm-x-45mm-OF-T-TPI/352821781036?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2057872.m2749.l2649), 1x [20 Ohm Resistor](https://www.segor.de/#Q=MF20R-1%2525&M=1), 1x [push button](https://www.segor.de/#/hilfe-zum-katalog). The price for all this parts is about 10€.

The pins for the wiring can be set or looked up in [tagger.h](https://github.com/open-laser-tag/tagger/blob/dev/src/tagger.h). Or look at the schematics below.
![Tagger schematics](https://github.com/open-laser-tag/tagger/blob/dev/docs/images/schematic.png)

![Tagger fritzing](https://github.com/open-laser-tag/tagger/blob/master/docs/images/tagger_Steckplatine.png)

## Features
#### Bluetooth (BLE)
When the tagger is connected via BLE, it is reporting the changes of its trigger and the incoming IR. With a connection the tagger doesn't send any infrared by its own and it doesn't control its LEDs. In BLE there can be different services, characteristics and properties on one device, which are building the BLE API. The tagger is running the following characteristics:

| UUID | Name | Properties | Description |
|---|---|---|---|
| 08dbb28a-ce2c-467a-9f12-4f15d574a220 | tagger service | all characteristics below | This service contains all tagger Characteristics. |
| 756ad6a4-2007-4dc4-9173-72dc7d6b2627 | trigger | notify | Get a notify about the trigger trigger status every time it changes. 1 for pressed, 0 for not pressed |
| a95980fb-4f18-4b2e-a258-81bf77575117 | IR receive | notify | Get a notify about the message when a message is incoming via infrared. |
| 8b91a0d2-5f7f-49cb-8939-4455d3d24b81 | IR send | write | Send a message to the tagger, to send it via infrared. |
| 60e44cef-5a43-407b-8d1a-bce02377dcfd | latency | notify | Get a notify about the time between the pressed trigger and the sent infrared. This is just for testing purpose. |
| 563c139f-3eda-4c88-9fc3-be987038fa6a | version | read | Read the tagger firmware version. |
| 7a4821c2-80f0-4eba-8070-d659d31e43de | led | write | Tell the tagger the color codes for the RGB LEDs: 0xNNRRGGBBNNRRGGBB... NN (uint8) is the position of the LED. RR (uint8) is red. GG (uint8) is green. BB (uint8) is blue. |
| eebc6352-2559-40f1-bda8-2715e7c07fbd | OTA | write | Write any data to this characteristic to turn on OTA mode.

#### Infrared (IR)
The main communication for a laser tag game is infrared. One IR LED [TSUS 5202](https://www.conrad.de/de/p/vishay-tsus-5202-cqw-13-ir-emitter-950-nm-15-5-mm-radial-bedrahtet-184551.html) is used to send IR. One IR receiver [TSOP31238](https://www.segor.de/#Q=TSOP31238&M=1) is used to detect infrared. Multiple receivers could be probably easily implemented. Both directions are using the [RMT peripheral](https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/peripherals/rmt.html). The infrared protocol can be chosen with `IR_PROTOCOL`, but it has to be the same on all devices. Currently it is NEC.

#### Over the air firmware update (OTA)
When the OTA signal is given via BLE, the tagger reboots in OTA mode. In OTA mode the tagger connects to a wifi given by `OTA_WIFI_SSID` and `OTA_WIWI_PASSWORD`. When no connetion is established in `TIME_WAITING_FOR_CONNECTION_IN_MS` (current: 10s), the tagger reboots again in normal mode. When the connection is established, the taggers firmware can be flashed via the local ip address through wifi.
More informations: https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/system/ota.html
When wifi connection is established but no new firmware is coming in, the tagger will stay in this status. When you want to get back to normal mode, just reboot the tagger (cut off power).

#### BLE-less mode
When the tagger is not connected via BLE, it does some things by its own to test some basic functionalities.

| LED | name | description |
| --- | --- | --- |
| 0 | tagger status | blinking red - choose your team with the trigger, red - tagger on and no BLE, blue - BLE established |
| 1 | player status | green - active, turns red for `PLAYER_DOWNTIME_IN_MS` (3s) when infrared message by another team is incoming |
| 2 | team | shows one of 7 team colors, the team differs the infrared message |
| 3 | trigger | blinking white when pressing trigger |

#### Debugging
The makros of [esp_log.h](https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/system/log.html) are used to send serial debug messages. When you open a serial monitor and connect the esp32 via USB you should see them. The build flag `DCORE_DEBUG_LEVEL` must be set for this.
When you got a JTAG adapter e.g. [jlink](https://www.ebay.de/itm/ARM7-ARM9-ARM11-J-link-V8-Emulator-Cortex-M3-IAR-STM32-JTAG-Schnittstelle-CYED/233389427917?hash=item3657180ccd:g:Ex4AAOSwYIxX-8DR) you can also do inline debugging with the [unified debugger of plattformio](https://docs.platformio.org/en/latest/plus/debugging.html).

## Licensing

This project is licensed under the terms of the GPL-3.0.


CHANGELOG.md
# v0.1.3
* add support for ota
* replace IRremote and own IR RMT module with IR32

# v0.1.2
* add support for APA102
* change debug logger from own usblog to esp_log.h

# v0.1.1
* add player status led
* add bt connection status led
* add player status: player is down when hit by ir for some seconds
* add teams to non-bt mode
* add a blinking led when shooting ir

# v0.1.0
* change IR hardware from YS-IRTM module to single IR LED

# v0.0.5
* add check whether message received via IR is valid (with last inverted byte)
* add plattformIO support
* drop support for more ir receiver: not working yet

# v0.0.4
* add sending error message 0xFFFFFF via IR when no BT connected
* add support for two more ir receiver
* add support for ws2812 led strip

# v0.0.3
* fixed bit order in received data from tsop ir reciever

# v0.0.2
* tagger firmware version readable via bluetooth
* new additional ir receiver with irremote.h lib
* add semaphores for handling communication rescources
* improve usb debug output

# v0.0.1
* first version of tagger firmware with trigger, infrared and bluetooth
* esp32
* freertos
* bluetooth low energy
* debounced trigger
* infrared with ir module YS-IRTM 
* latency between pressed trigger and ir shot is sent via bluetooth


src/tagger.h
/**
 * @file globals.h
 * @author gldhnchn (gldhnchn@posteo.de)
 * @brief global variables
 * @date 2019-01-23
 * 
 */
#ifndef tagger_h
#define tagger_h

/* ++++++ */
/* config */
// these configs can safely be editted for each tagger seperatly

/* RGB LEDs */
//choose one LED type
//#define LED_TYPE_APA102
#define LED_TYPE_NEOPIXEL
#define LED_CLOCK_PIN 18 //only used for spi leds like APA102
#define LED_DATA_PIN 19
#define NUM_LEDS 4
#define LED_OVERALL_BRIGHTNESS 20 //scale 0 to 255
//FASTLED_ESP32_I2S is defined to tell FastLED to use I2S driver instead of RMT. Reference for this option in FastLED/platforms/esp/32/clockless_i2s_esp32.h. There is a conflict when using RMT with both LED and IR, although it should be possible according to the options in FastLED/platforms/esp/32/clockless_rmt_esp32.h. Maybe there is a connection to this issue: https://github.com/FastLED/FastLED/issues/835
#define FASTLED_ESP32_I2S true
#define LED_INDEX_BT 0
#define LED_INDEX_PLAYER_STATUS 1
#define LED_INDEX_TEAM 2
#define LED_INDEX_SHOOT 3
#define COLOR_BT_CONNECTION_ON 0x0000FF
#define COLOR_BT_CONNECTION_OFF 0xFF0000
#define COLOR_PLAYER_STATUS_ON 0x00FF00
#define COLOR_PLAYER_STATUS_OFF 0xFF0000
#define ONBOARDLED_PIN 2

/* Infra Red */
#define IR_RECV_FRONT_PIN 27
#define IR_RECV_RIGHT_PIN 26
#define IR_RECV_LEFT_PIN 25
#define IR_RMT_TX_CHANNEL RMT_CHANNEL_0 /*!< RMT channel for transmitter */
#define IR_RMT_RX_FRONT_CHANNEL RMT_CHANNEL_1
#define IR_RMT_TX_GPIO_NUM GPIO_NUM_22 /*!< GPIO number for transmitter signal */

/* Trigger Button */
#define PIN_TRIGGER 21
#define DEBOUNCETIME 10

/* OTA */
#define OTA_WIFI_SSID "wlanssid"
#define OTA_WIWI_PASSWORD "wlanpassword"
#define EEPROM_ADDR_OTA 0
#define EEPROM_SIZE 1

/* Timings */
#define PLAYER_DOWNTIME_IN_MS 3000
#define TEAM_SELECTION_TIME_IN_MS 5000

/* ++++++++ */
/* includes */
#include <Arduino.h>
#include <BLEDevice.h>
#include <BLE2902.h>
#include "misc.h"
#include "ir_and_trigger.h"
#include "ble.h"
#include <FastLED.h>
#include <esp_log.h>
#include "ota.h"
#include <EEPROM.h>
#include "IR32/src/IRSend.h"
#include "IR32/src/IRRecv.h"

/* git version */
#if __has_include("git_info.h")
#include "git_info.h"
#else
#define GIT_TAG "v0.1.3"
#endif

/* ++++++ */
/* macros */
// Don't edit this section. It has to be equal on all taggers.
// See the following for generating UUIDs:
// https://www.uuidgenerator.net/
#define SERVICE_UUID "08dbb28a-ce2c-467a-9f12-4f15d574a220"
#define CHARACTERISTIC_TRIGGER_UUID "756ad6a4-2007-4dc4-9173-72dc7d6b2627"
#define CHARACTERISTIC_IR_RECEIVE_UUID "a95980fb-4f18-4b2e-a258-81bf77575117"
#define CHARACTERISTIC_IR_SEND_UUID "8b91a0d2-5f7f-49cb-8939-4455d3d24b81"
#define CHARACTERISTIC_LATENCY_UUID "60e44cef-5a43-407b-8d1a-bce02377dcfd"
#define CHARACTERISTIC_VERSION_UUID "563c139f-3eda-4c88-9fc3-be987038fa6a"
#define CHARACTERISTIC_LED_UUID "7a4821c2-80f0-4eba-8070-d659d31e43de"
#define CHARACTERISTIC_OTA_UUID "eebc6352-2559-40f1-bda8-2715e7c07fbd"
#define IR_PROTOCOL "NEC"              //choose from timing groups lib/IR32/src/IR32.h

/* global variables*/
extern const uint32_t color_team[];
extern uint32_t last_time_button_pressed_timestamp,
    latency,
    last_bounce_time;
extern const uint32_t ir_msg[];
extern uint16_t count_trigger_interrupts,
    msg_nr;
extern uint8_t team;
extern bool team_selection;
extern portMUX_TYPE mux;
extern TaskHandle_t xHandle_handle_ir,
    xHandle_refresh_trigger_status,
    xHandle_send_latency,
    xHandle_handle_player_status;
extern SemaphoreHandle_t xMutex_BT;
extern BLECharacteristic *trigger_char,
    *ir_receive_char,
    *ir_send_char,
    *latency_char,
    *version_char,
    *led_char,
    *ota_char;
extern IRSend ir_led;
extern IRRecv ir_recv_front;
extern Led led;
extern Button trigger;
extern CRGB leds[NUM_LEDS];
extern bool player_is_on;

/* function declerations */
void setup();
void loop();
void create_tasks();
void send_latency(void *parameter);
void init_mutex();
#endif // tagger_h
